File: backend/UnderSea.Bll/Dtos/Upgrade/UpgradeDto.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace UnderSea.Bll.Dtos
{
    public class UpgradeDto
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public bool DoesExist { get; set; }
        public bool IsUnderConstruction { get; set; }
        public int RemainingTime { get; set; }
        public string ImageUrl { get; set; }
        public ICollection<EffectDto> Effects { get; set; }
    }
}

File: frontend/flutter/undersea/lib/models/response/upgrade_dto.dart
import 'package:json_annotation/json_annotation.dart';

import 'effect_dto.dart';

part 'upgrade_dto.g.dart';

@JsonSerializable()
class UpgradeDto {
  int id;
  String name;
  bool doesExist;
  bool isUnderConstruction;
  int remainingTime;
  String? imageUrl;
  List<EffectDto>? effects;

  UpgradeDto(
      {this.effects,
      required this.id,
      required this.doesExist,
      required this.isUnderConstruction,
      required this.name,
      required this.remainingTime,
      this.imageUrl});

  factory UpgradeDto.fromJson(Map<String, dynamic> json) =>
      _$UpgradeDtoFromJson(json);

  /// `toJson` is the convention for a class to declare support for serialization
  /// to JSON. The implementation simply calls the private, generated
  /// helper method `_$UserToJson`.
  Map<String, dynamic> toJson() => _$UpgradeDtoToJson(this);
}
